src/main/scala/viper/silver/plugin/standard/adt/encoding/AdtEncoder.scala

```diff
@@ -427,6 +441,171 @@ class AdtEncoder(val program: Program) extends AdtNameManager {
+    AnonymousDomainAxiom(forall)(domain.pos, domain.info, getWellFoundedDomainName(domain.name), domain.errT)
+  }
+
+  /**
+    * Generates an axiom that expresses that all values of the current ADT are bounded:
+    * forall x: ADT :: { bounded(x) } bounded(x)
+    * This is justified by the fact that Viper ADTs are recursive types and are always finite.
+    */
+  private def generateBoundedAxiom(domain: Domain): AnonymousDomainAxiom = {
+    val domainType = DomainType(domain, defaultTypeVarsFromDomain(domain))
+    val param = LocalVarDecl("x", domainType)()
+    val boundedApp = DomainFuncApp(
+      getBoundedFunctionName,
+      Seq(param.localVar),
+      Map(TypeVar("T") -> domainType)
+    )(domain.pos, domain.info, Bool, getWellFoundedOrderDeclarationDomainName, domain.errT)
+    val trigger = Trigger(Seq(boundedApp))(domain.pos, domain.info, domain.errT)
+    val forall = Forall(Seq(param), Seq(trigger), boundedApp)(domain.pos, domain.info, domain.errT)
+    AnonymousDomainAxiom(forall)(domain.pos, domain.info, getWellFoundedDomainName(domain.name), domain.errT)
+  }
+
+  /**
+    * Generates an axiom for the given constructor that expresses that all values of the current ADT contained
+    * inside an ADT value constructed using said constructor are smaller than the ADT value itself.
+    * E.g., for List { Nil() Cons(i: Int, l: List) }:
+    * forall i: Int, l: List :: { Cons(i, l) } decreases(l, Cons(i, l))
+    * Also takes into account values that may be contained through constructors of other ADT types (in cases of mutually
+    * recursive ADT definitions).
+    */
+  private def generateDecreasesAxiom(domain: Domain)(ac: AdtConstructor): AnonymousDomainAxiom = {
+    assert(domain.name == ac.adtName, "AdtEncoder: An error in the ADT encoding occurred.")
+
+    val localVars = ac.formalArgs.map { lv =>
+      lv.typ match {
+        case a: AdtType => localVarTFromType(encodeAdtTypeAsDomainType(a), Some(lv.name))(ac.pos, ac.info, ac.errT)
+        case d => localVarTFromType(d, Some(lv.name))(ac.pos, ac.info, ac.errT)
+      }
+    }
+
+    val decreases = if (ac.formalArgs.isEmpty) {
+      Nil
+    } else {
+      val localVarDecl = ac.formalArgs.collect { case l: LocalVarDecl => l }
+
+      assert(localVarDecl.size == localVars.size, "AdtEncoder: An error in the ADT encoding occurred.")
+
+      /**
+        * Given a variable currentVar that represents an argument of the current ADT constructor, if the variable's
+        * type is an ADT type, recursively looks for values of the original ADT type, either in the variable itself
+        * or in all constructors of its type.
+        * E.g., if ac is a constructor for ADT1, then:
+        *  - if the type of currentVar is ADT1, then we have already found a value of the original ADT type
+        *  - if the type of the variable is ADT2, and ADT2 has a constructor C(T1, ADT1), then we have found a value
+        *    of the original type inside this constructor.
```

Comment:
Another approach for the encoding would be to generate a rank function for every ADT. We could axiomatize `decreasing(x, y) = rank(x) < rank(y)` and, for every "base case" `z`, have `rank(z) == 0`. This is a bit cumbersome, but the advantage is that we do not need to recurse into ADT definitions to find mutually recursive uses of an ADT. Instead, we could emit the following axiom, for every constructor `C(t: T')` of ADT type T, where `T'` is allowed to be any ADT type including `T`:
```
rank(t) < rank(C(t))
```

At any rate, this solution is also good.

Reply:
Oh yeah that's a neat idea.

Reply:
But I'll merge it as is, this should work just as well and I don't feel like rewriting it.

Reply:
sounds good 😅